pkg/controller/controller.go

```diff
@@ -118,29 +128,86 @@ type controller struct {
 	brokerLister              listers.BrokerLister
 	serviceClassLister        listers.ServiceClassLister
 	instanceLister            listers.InstanceLister
+	bindingLister             listers.BindingLister
 	brokerRelistInterval      time.Duration
 	enableOSBAPIContextProfle bool
 	recorder                  record.EventRecorder
+	brokerQueue               workqueue.RateLimitingInterface
+	serviceClassQueue         workqueue.RateLimitingInterface
+	instanceQueue             workqueue.RateLimitingInterface
+	bindingQueue              workqueue.RateLimitingInterface
 }
 
 // Run runs the controller until the given stop channel can be read from.
-func (c *controller) Run(stopCh <-chan struct{}) {
+func (c *controller) Run(workers int, stopCh <-chan struct{}) {
 	defer runtimeutil.HandleCrash()
+
 	glog.Info("Starting service-catalog controller")
 
+	for i := 0; i < workers; i++ {
+		go wait.Until(worker(c.brokerQueue, "Broker", c.reconcileBrokerKey), time.Second, stopCh)
+		go wait.Until(worker(c.serviceClassQueue, "ServiceClass", c.reconcileServiceClassKey), time.Second, stopCh)
+		go wait.Until(worker(c.instanceQueue, "Instance", c.reconcileInstanceKey), time.Second, stopCh)
+		go wait.Until(worker(c.bindingQueue, "Binding", c.reconcileBindingKey), time.Second, stopCh)
+	}
+
 	<-stopCh
 	glog.Info("Shutting down service-catalog controller")
+
+	c.brokerQueue.ShutDown()
+	c.serviceClassQueue.ShutDown()
+	c.instanceQueue.ShutDown()
+	c.bindingQueue.ShutDown()
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the reconciler is never invoked concurrently with the same key.
+func worker(queue workqueue.RateLimitingInterface, resourceType string, reconciler func(key string) error) func() {
+	return func() {
+		exit := false
+		for !exit {
+			exit = func() bool {
+				key, quit := queue.Get()
+				if quit {
+					return true
+				}
+				defer queue.Done(key)
+
+				err := reconciler(key.(string))
+				if err == nil {
```

Comment:
1) unless there is a programming error, yes
2) we use err below so the scope needs to extend outside the if block